### ISSUE TITLE
Cluster blocked after pause minority

### ISSUE THREAD

**Issue by dcorbacho** [rabbitmq/rabbitmq-server#676]
# Test

Rabbit cluster with 3 Ubuntu machines (rabbit04, rabbit05, rabbit06) running 3.6.1.

While running the tests to reproduce https://github.com/rabbitmq/rabbitmq-server/issues/545, the cluster got stuck for over 48h and it doesn't unblocks.

Logs are available [here](https://drive.google.com/folderview?id=0B7YDkFQ4mjPmbG12N3dENnYxdmM&usp=sharing). The startup_log contains extra dbg logs

## rabbit04
Is blocked in `rabbit:stop()` awaiting for `rabbit_boot` to finish. The boot process is waiting for `mnesia:change_config/2` which is at the same time stuck in `mnesia_controller:connect_nodes/1`.
The controller is awaiting in this receive clause: https://github.com/erlang/otp/blob/maint/lib/mnesia/src/mnesia_controller.erl#L460 for the schema to be merged in `try_merge_schema`.

Mnesia seems to be blocked in a transaction aborted as `cyclic` over the `rabbit_exchange` table. Which in fact, doesn't appear in any of the mnesia locks. 

```
2016-03-03 08:18:51 [{pid,<5588.4805.0>},
                     {registered_name,[]},
                     {current_stacktrace,
                         [{timer,sleep,1,[{file,"timer.erl"},{line,153}]},
                          {rabbit,await_startup,1,
                              [{file,"src/rabbit.erl"},{line,378}]},
                          {rabbit,stop,0,[{file,"src/rabbit.erl"},{line,323}]},
                          {rabbit,stop_and_halt,0,
                              [{file,"src/rabbit.erl"},{line,332}]},
                          {rpc,'-handle_call_call/6-fun-0-',5,
                              [{file,"rpc.erl"},{line,206}]}]},
                     {initial_call,{erlang,apply,2}},
                     {message_queue_len,0},
                     {links,[]},
                     {monitors,[]},
                     {monitored_by,[<5588.14.0>]},
                     {heap_size,233}]
2016-03-03 08:18:51 [{pid,<5588.4503.0>},
                     {registered_name,rabbit_outside_app_process},
                     {current_stacktrace,
                         [{gen,do_call,4,[{file,"gen.erl"},{line,168}]},
                          {gen_server,call,3,
                              [{file,"gen_server.erl"},{line,208}]},
                          {app_utils,'-manage_applications/6-fun-0-',6,
                              [{file,"src/app_utils.erl"},{line,118}]},
                          {lists,foldl,3,[{file,"lists.erl"},{line,1262}]},
                          {app_utils,manage_applications,6,
                              [{file,"src/app_utils.erl"},{line,117}]},
                          {rabbit,start_apps,1,
                              [{file,"src/rabbit.erl"},{line,346}]},
                          {rabbit,broker_start,0,
                              [{file,"src/rabbit.erl"},{line,286}]},
                          {rabbit,start_it,1,
                              [{file,"src/rabbit.erl"},{line,303}]}]},
                     {initial_call,{erlang,apply,2}},
                     {message_queue_len,0},
                     {links,[<5588.4537.0>]},
                     {monitors,[{process,<5588.7.0>}]},
                     {monitored_by,[]},
                     {heap_size,6772}]
2016-03-03 08:18:51 [{pid,<5588.4580.0>},
                     {registered_name,rabbit},
                     {current_stacktrace,
                         [{mnesia_controller,connect_nodes,2,
                              [{file,"mnesia_controller.erl"},{line,460}]},
                          {rabbit_mnesia,change_extra_db_nodes,2,
                              [{file,"src/rabbit_mnesia.erl"},{line,758}]},
                          {rabbit_mnesia,init_db,3,
                              [{file,"src/rabbit_mnesia.erl"},{line,447}]},
                          {rabbit_mnesia,init_db_and_upgrade,3,
                              [{file,"src/rabbit_mnesia.erl"},{line,478}]},
                          {rabbit_mnesia,init,0,
                              [{file,"src/rabbit_mnesia.erl"},{line,110}]},
                          {rabbit_boot_steps,'-run_step/2-lc$^1/1-1-',1,
                              [{file,"src/rabbit_boot_steps.erl"},{line,49}]},
                          {rabbit_boot_steps,run_step,2,
                              [{file,"src/rabbit_boot_steps.erl"},{line,49}]},
                          {rabbit_boot_steps,'-run_boot_steps/1-lc$^0/1-0-',
                              1,
                              [{file,"src/rabbit_boot_steps.erl"},
                               {line,26}]}]},
                     {initial_call,{application_master,start_it,4}},
                     {message_queue_len,1},
                     {links,[<5588.4581.0>,<5588.4610.0>,<5588.4579.0>]},
                     {monitors,[]},
                     {monitored_by,[<5588.4604.0>]},
                     {heap_size,10958}] 
```

## rabbit05

Is stuck with delegate calls and in-flight transactions. Also, several errors of channel terminate timeout.

```
=CRASH REPORT==== 2-Mar-2016::14:48:30 ===
  crasher:
    initial call: gen:init_it/6
    pid: <0.5206.0>
    registered_name: []
    exception exit: {{badmatch,{error,{channel_operation_timeout,5000}}},
                     [{rabbit_channel,terminate,2,
                                      [{file,"src/rabbit_channel.erl"},
                                       {line,614}]},
                      {gen_server2,terminate,3,
                                   [{file,"src/gen_server2.erl"},{line,1146}]},
                      {proc_lib,wake_up,3,
                                [{file,"proc_lib.erl"},{line,250}]}]}
      in function  gen_server2:terminate/3 (src/gen_server2.erl, line 1149)
    ancestors: [<0.5203.0>,<0.5199.0>,<0.5188.0>,<0.5187.0>,<0.665.0>,
                  <0.664.0>,<0.663.0>,rabbit_sup,<0.120.0>]
    messages: [{'DOWN',#Ref<0.0.2.52544>,process,<7847.3102.0>,normal},
                  {'DOWN',#Ref<0.0.2.52545>,process,<7846.9755.0>,normal}]
    links: []
    dictionary: [{{xtype_to_module,fanout},rabbit_exchange_type_fanout},
                  {msg_size_for_gc,24000},
                  {{credit_from,<7846.9755.0>},200},
                  {{credit_from,<0.5214.0>},200},
                  {process_name,
                      {rabbit_channel,
                          {<<"10.100.0.121:46213 -> 10.100.0.85:5672">>,1}}},
                  {delegate,delegate_4},
                  {guid_secure,
                      {{3,'rabbit@t-srv-rabbit05',#Ref<0.0.1.53952>},1}},
                  {channel_operation_timeout,5000},
                  {{credit_from,<7847.3102.0>},200},
                  {credit_flow_default_credit,{200,50}},
                  {{credit_to,<0.5189.0>},50},
                  {guid,{{3499031759,2980528650,367330770,1247722968},799}}]
    trap_exit: true
    status: running
    heap_size: 4185
    stack_size: 27
    reductions: 248033
  neighbours:
```

There seems to be a cyclic transaction error in mnesia, which causes the transaction to be restarted. It is executing a function from rabbit_amqqueue.

```
(<0.4753.0>) call mnesia:select_lock({tid,2243,<0.4753.0>},{tidstore,5267528,[],1},read,[{'$1',[],['$1']}],rabbit_queue)
(<0.4753.0>) call mnesia:lock_table({tid,2243,<0.4753.0>},{tidstore,5267528,[],1},rabbit_queue,read)
(<0.4753.0>) exception_from {mnesia,lock_table,4} {exit,{aborted,{cyclic,'rabbit@t-srv-rabbit05',
                       {rabbit_queue,'______WHOLETABLE_____'},
                       read,read,
                       {tid,1893,<0.4603.0>}}}}
(<0.4753.0>) call mnesia_tm:restart(mnesia,{tid,2243,<0.4753.0>},{tidstore,5267528,[],1},#Fun<rabbit_amqqueue.33.41098263>,[],758,infinity,sync,{cyclic,'rabbit@t-srv-rabbit05',
        {rabbit_queue,'______WHOLETABLE_____'},
        read,read,
        {tid,1893,<0.4603.0>}})
```
The transaction which might being handled is https://github.com/rabbitmq/rabbitmq-common/blob/stable/src/rabbit_amqqueue.erl#L891, as the maybe stuck output shows a process blocked handling the down of a node and the dbg output over mnesia shows also qlc queries. In this case, there are write and read locks over the `rabbit_queue` table.

# rabbit06

This node seems stuck trying to sync the mirrored queues:
```
2016-03-03 08:38:29 [{pid,<5588.1588.0>},
                     {registered_name,[]},
                     {current_stacktrace,
                         [{rabbit_mirror_queue_sync,master_go,8,
                              [{file,"src/rabbit_mirror_queue_sync.erl"},
                               {line,94}]},
                          {rabbit_mirror_queue_master,sync_mirrors,3,
                              [{file,"src/rabbit_mirror_queue_master.erl"},
                               {line,159}]},
                          {rabbit_amqqueue_process,handle_call,3,
                              [{file,"src/rabbit_amqqueue_process.erl"},
                               {line,1112}]},
                          {gen_server2,handle_msg,2,
                              [{file,"src/gen_server2.erl"},{line,1028}]},
                          {proc_lib,wake_up,3,
                              [{file,"proc_lib.erl"},{line,250}]}]},
                     {initial_call,{proc_lib,init_p,5}},
                     {message_queue_len,520},
                     {links,
                         [<5588.1587.0>,<5588.1832.0>,<5588.5645.0>,
                          #Port<5588.8835>]},
                     {monitors,[]},
                     {monitored_by,
                         [<5588.931.0>,<5588.5644.0>,<5588.923.0>,
                          <5588.934.0>,<5588.822.0>]},
                     {heap_size,514838}]
```

There is another transaction permanently restarted in mnesia, originates from `mnesia_loader` as in rabbit04

```
(<0.5754.0>) exception_from {mnesia_tm,apply_fun,3} {exit,{aborted,{cyclic,'rabbit@t-srv-rabbit05',
                       {rabbit_durable_queue,'______WHOLETABLE_____'},
                       write,write,
                       {tid,1893,<21740.4603.0>}}}}
(<0.5754.0>) call mnesia_tm:check_exit(#Fun<mnesia_loader.4.89306253>,[],761,infinity,{aborted,{cyclic,'rabbit@t-srv-rabbit05',
                 {rabbit_durable_queue,'______WHOLETABLE_____'},
                 write,write,
                 {tid,1893,<21740.4603.0>}}},async)
(<0.5754.0>) call mnesia_tm:maybe_restart(#Fun<mnesia_loader.4.89306253>,[],761,infinity,async,{cyclic,'rabbit@t-srv-rabbit05',
        {rabbit_durable_queue,'______WHOLETABLE_____'},
        write,write,
        {tid,1893,<21740.4603.0>}})
(<0.5754.0>) call mnesia_tm:try_again(infinity)
(<0.5754.0>) returned from mnesia_tm:try_again/1 -> yes
(<0.5754.0>) call mnesia_tm:restart(mnesia,{tid,2329,<0.5754.0>},{tidstore,7417936,[],1},#Fun<mnesia_loader.4.89306253>,[],761,infinity,async,{cyclic,'rabbit@t-srv-rabbit05',
        {rabbit_durable_queue,'______WHOLETABLE_____'},
        write,write,
        {tid,1893,<21740.4603.0>}})
```

**Comment by dcorbacho** [comment]
The underlying problem is in fact a mnesia bug triggered in node rabbit05.

The maybe_stuck output on that node shows that a mnesia transaction is always blocked in `mnesia_tm:rec_acc_pre_commit`, thus other transactions trying to access the locked tables are permanently restarted (`mnesia_tm:restart`):

```
2016-03-03 08:31:42 [{pid,<5588.4603.0>},
                     {registered_name,[]},
                     {current_stacktrace,
                         [{mnesia_tm,rec_acc_pre_commit,8,
                              [{file,"mnesia_tm.erl"},{line,1550}]},
                          {mnesia_tm,apply_fun,3,
                              [{file,"mnesia_tm.erl"},{line,834}]},
                          {mnesia_tm,execute_transaction,5,
                              [{file,"mnesia_tm.erl"},{line,808}]},
                          {rabbit_misc,
                              '-execute_mnesia_transaction/1-fun-0-',1,
                              [{file,"src/rabbit_misc.erl"},{line,534}]},
                          {worker_pool_worker,'-run/2-fun-0-',3,
                              [{file,"src/worker_pool_worker.erl"},
                               {line,81}]}]},
```

The mnesia process is waiting for a 'mnesia_down' message after the network partition is triggered, but that message never arrives https://github.com/erlang/otp/blob/maint/lib/mnesia/src/mnesia_tm.erl#L1548

This bug has been reported and debugged with help of the OTP team (thanks @dgud!), which provided us with a patch for it. The patch is going through a last round of testing with RabbitMQ, and will be linked here once it is publicly available. 

As a curiosity, the bug is triggered when using mirrored queues: `rabbit_mirror_queue_slave:handle_go` and `rabbit_mirror_queue_slave:record_synchronised` are the functions triggering the issue. Both `rabbit_mirror_queue_slave` functions run almost the same mnesia transaction:

1. Read a queue from `rabbit_queue` table
2. Call `rabbit_mirror_queue_misc:store_updated_slaves`

**Comment by noahhaon** [comment]
I am confused, does the resolution for this issue require an OTP upgrade, or will RabbitMQ be mitigating the Mnesia bug in a later release?

**Comment by dcorbacho** [comment]
Hi @noahhaon, this issue will require an OTP upgrade. We will publish the details here whenever the OTP patch is available.

**Comment by noahhaon** [comment]
@dcorbacho thank you for the quick response!  I assume it is not possible/practical to mitigate this in the RabbitMQ application?  We have issues upgrading to Erlang 18 due to openssl incompatibilities.

**Comment by dcorbacho** [comment]
@noahhaon we don't have any workaround in RabbitMQ. 

If upgrade is not a possibility, you might want to evaluate the option of backporting the Mnesia patch and build your own Erlang release. It is always a bit delicate, but there have been few changes in Mnesia lately.

**Comment by noahhaon** [comment]
@dcorbacho Probably better to resolve our OpenSSL issues to get the other Mnesia fixes then.  Mnesia has been a sore point for us at scale, so it would likely be worth the pain of an upgrade.  Thanks!

**Comment by dcorbacho** [comment]
The internal OTP bug tracker number is OTP-13423, the patch will be hopefully available this week.

**Comment by dcorbacho** [comment]
OTP patch release 18.3.1 is out https://github.com/erlang/otp/tree/OTP-18.3.1

```
Patch Package:           OTP 18.3.1
Git Tag:                 OTP-18.3.1
Date:                    2016-04-04
Trouble Report Id:       OTP-13417, OTP-13418, OTP-13419, OTP-13420,
                         OTP-13423, OTP-13424, OTP-13446, OTP-13452
Seq num:
System:                  OTP
Release:                 18
Application:             erts-7.3.1, inets-6.2.1, mnesia-4.13.4
Predecessor:             OTP 18.3

 Check out the git tag OTP-18.3.1, and build a full OTP system
 including documentation. Apply one or more applications from this
 build as patches to your installation using the 'otp_patch_apply'
 tool. For information on install requirements, see descriptions for
 each application version below.

 ---------------------------------------------------------------------
 --- erts-7.3.1 ------------------------------------------------------
 ---------------------------------------------------------------------

 The erts-7.3.1 application can be applied independently of other
 applications on a full OTP 18 installation.

 --- Fixed Bugs and Malfunctions ---

  OTP-13418    Application(s): erts

               process_info(Pid, last_calls) did not work for Pid /=
               self().


  OTP-13419    Application(s): erts

               Make sure to create a crash dump when running out of
               memory. This was accidentally removed in the erts-7.3
               release.


  OTP-13420    Application(s): erts

               Schedulers could be woken by a premature timeout on
               Linux. This premature wakeup was however harmless.


  OTP-13424    Application(s): erts
               Related Id(s): OTP-10336

               A process communicating with a port via one of the
               erlang:port_* BIFs could potentially end up in an
               inconsistent state if the port terminated during the
               communication. When this occurred the process could
               later block in a receive even though it had messages
               matching in its message queue.

               This bug was introduced in erts version 5.10 (OTP
               R16A).


  OTP-13446    Application(s): erts

               The reference count of a process structure could under
               rare circumstances be erroneously managed. When this
               happened invalid memory accesses occurred.


  OTP-13452    Application(s): erts

               Fix race between process_flag(trap_exit,true) and a
               received exit signal.

               A process could terminate due to exit signal even
               though process_flag(trap_exit,true) had returned. A
               very specific timing between call to process_flag/2 and
               exit signal from another scheduler was required for
               this to happen.


 Full runtime dependencies of erts-7.3.1: kernel-4.0, sasl-2.4,
 stdlib-2.5


 ---------------------------------------------------------------------
 --- inets-6.2.1 -----------------------------------------------------
 ---------------------------------------------------------------------

 The inets-6.2.1 application can be applied independently of other
 applications on a full OTP 18 installation.

 --- Fixed Bugs and Malfunctions ---

  OTP-13417    Application(s): inets

               Mend ipv6_host_with_brackets option in httpc


 Full runtime dependencies of inets-6.2.1: erts-6.0, kernel-3.0,
 mnesia-4.12, runtime_tools-1.8.14, ssl-5.3.4, stdlib-2.0


 ---------------------------------------------------------------------
 --- mnesia-4.13.4 ---------------------------------------------------
 ---------------------------------------------------------------------

 The mnesia-4.13.4 application can be applied independently of other
 applications on a full OTP 18 installation.

 --- Fixed Bugs and Malfunctions ---

  OTP-13423    Application(s): mnesia

               Mnesia transactions could hang while waiting on a
               response from a node who had stopped.


 Full runtime dependencies of mnesia-4.13.4: erts-7.0, kernel-3.0,
 stdlib-2.0


 ---------------------------------------------------------------------
 --- Thanks to -------------------------------------------------------
 ---------------------------------------------------------------------

 Andrey Mayorov, Lukas Larsson
```

**Comment by michaelklishin** [comment]
Closing this as the root cause is supposed to be fixed in OTP 18.3.1. That said, we have a different issue discovered as part of this: #714, will continue there.
